### PR TITLE
Switch from JSON to MsgPack data transport

### PR DESCRIPTION
**Pull request by mavam**
This PR changes how we receive data from Fluent Bit. Previously, we told Fluent Bit that we want to receive data in JSON in the library callback. Unfortunately this format isn't robust for large messages. By "robust", we mean that we could get invalid JSON due to the following Fluent Bit code in `out_lib.c:184`:

```c
len = snprintf(out_buf, out_size, "[%f,%s]",
               flb_time_to_double(&tm),
               buf);
```

Here, `buf` holds the Fluent Bit message and can contain cropped JSON that then gets wrapped into the common array-style framing.

To avoid running into this problem we now go through a different code path in Fluent Bit, avoiding JSON conversion alltogether. By switching the format of the `out_lib` plugin to MsgPack, we now have direct access to the internal representation of Fluent Bit that does not get cropped.

A beneficial side effect of this change is that we should have now higher throughput as well due to the absence of an extra JSON roundtrip.

### CODE REVIEW

**Comment by mavam** [comment]
@tobim could I have some assistance on packaging? It looks like we need to depend on the MsgPack C library now, according to the Docker CI failure:

```
1291.1 /tmp/tenzir/plugins/fluent-bit/src/plugin.cpp:24:10: fatal error: msgpack.h: No such file or directory
[1291](https://github.com/tenzir/tenzir/actions/runs/7354368856/job/20021529157?pr=3770#step:10:1292).1    24 | #include <msgpack.h>
1291.1       |          ^~~~~~~~~~~
```

Fluent Bit uses MsgPack internally. It looks like they don't expose that dependency publicly and we don't get it transitively.

**Review by Dakostu — APPROVED** [review]
OK - I can also confirm the performance boost.

Some comments remain.

File: plugins/fluent-bit/src/plugin.cpp [hunk 1]
```diff
@@ -39,6 +81,130 @@ namespace tenzir::plugins::fluentbit {
...
+}
+
+auto to_bin(const msgpack_object& object) {
+  return std::span<const std::byte>{
+    reinterpret_cast<const std::byte*>(object.via.bin.ptr),
+    size_t{object.via.bin.size}};
+}
+
+/// A MsgPack object.
+auto visit(auto f, const msgpack_object& object) {
```
Comment (Dakostu) [inline plugins/fluent-bit/src/plugin.cpp:108]: `default` case missing here - intended?

File: changelog/next/changes/3770.md [hunk 2]
```diff
@@ -0,0 +1,6 @@
+The `fluent-bit` source operator changed how it receives data from Fluent Bit.
+The operator no longer requests JSON conversion from Fluent Bit prior to
+processing an event, but instead directly processes the MsgPack data that Fluent
+Bit uses internally. Using MsgPack increases robustness of event delivery, as we
+found corner cases where Fluent Bit crops JSON message objects. The change also
+increases throughput due to the absence of an extra fromat conversion.
```
[the remark this answers is not in the document]
Comment (mavam) [reply changelog/next/changes/3770.md:6]: Reads definitely smoother, thanks!

File: plugins/fluent-bit/src/plugin.cpp (hunk 1 above)
Comment (mavam) [reply plugins/fluent-bit/src/plugin.cpp:108]: Yes. In this case the compiler will warn me if I forgot an enum.

**Comment by mavam** [comment]
> OK - I can also confirm the performance boost.

Do you have any numbers? I tried this with a `Release` build:

```bash
cat /tmp/m57/Suricata/eve.json | tenzir 'fluent-bit stdin | head 10M | discard'
```

I get:

- this branch: **10 seconds**
- main branch: **41 seconds**

So I'm seeing a **4x** performance improvement.

Runtime from `--dump-metrics`:

```
operator #1 (fluent-bit)
  total: 10.41s
  scheduled: 10.0s (96.05%)
  processing: 9.86s (94.67%)
  runs: 1658909 (0.00% processing / 0.00% input / 0.00% output)
  outbound:
    events: 1658700 at a rate of 159269.33/s
    bytes: 3147068950 at a rate of 302183367.44/s (estimate)
    batches: 27 (61433.33 events/batch)
```

Main:

```
operator #1 (fluent-bit)
  total: 42.78s
  scheduled: 42.75s (99.91%)
  processing: 42.69s (99.78%)
  runs: 93763 (96.06% processing / 0.00% input / 96.06% output)
  outbound:
    events: 2149357 at a rate of 50237.24/s
    bytes: 1521111771 at a rate of 35553169.67/s (estimate)
    batches: 90065 (23.86 events/batch)
```

EPS went from **50k** to **150k*. (**3x**)

**Comment by Dakostu** [comment]
@mavam Another small experiment with two `Release` builds and a ~500 MB `eve.json` file.

PR build:
```
cat /home/dakostu/Downloads/Suricata/eve.json | time ./tenzir --plugins=fluent-bit "fluent-bit stdin | head 10M | discard"

./tenzir --plugins=fluent-bit "fluent-bit stdin | head 10M | discard" 
17,68s user 
0,45s system 
325% cpu
5,561 total
```


`main` build:

```
cat /home/dakostu/Downloads/Suricata/eve.json | time ./tenzir --plugins=fluent-bit "fluent-bit stdin | head 10M | discard"

./tenzir --plugins=fluent-bit "fluent-bit stdin | head 10M | discard" 
133,09s user
1,42s system
310% cpu
43,354 total
```

In this case, there is an ~8x performance improvement.
This performance difference looks to be larger, the bigger the dataset is.

`dump-metrics` stats for PR build:
```
operator #1 (fluent-bit)
  total: 4.94s
  scheduled: 4.83s (97.79%)
  processing: 4.78s (96.63%)
  runs: 648419 (0.00% processing / 0.00% input / 0.00% output)
  outbound:
    events: 648232 at a rate of 131115.33/s
    bytes: 834211835 at a rate of 168732743.81/s (estimate)
    batches: 10 (64823.20 events/batch)
```

For `main` build:

``` 
operator #1 (fluent-bit)
  total: 39.23s
  scheduled: 39.21s (99.97%)
  processing: 39.19s (99.91%)
  runs: 36046 (97.17% processing / 0.00% input / 97.17% output)
  outbound:
    events: 902967 at a rate of 23019.58/s
    bytes: 510210972 at a rate of 13006945.45/s (estimate)
    batches: 35026 (25.78 events/batch)
```

EPS rate improvement is around x5 here.

**Pull request merged**